File: app/src/main/java/com/nata/volumecontrol/MainActivity.java
package com.nata.volumecontrol;

import android.content.Intent;
import android.os.Bundle;
import android.provider.Settings;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.firebase.jobdispatcher.FirebaseJobDispatcher;
import com.firebase.jobdispatcher.GooglePlayDriver;
import com.firebase.jobdispatcher.Job;
import com.firebase.jobdispatcher.Lifetime;
import com.firebase.jobdispatcher.Trigger;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();

    private ProgressBar progressCurrent;
    private Button btnRefresh;
    private Button btnFlush;
    private TextView tvTotalPlayed;
    private TextView tvMaxDuration;

    private VolumeControlService volumeControlService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        volumeControlService = new VolumeControlService();

        progressCurrent = findViewById(R.id.progress_current);
        btnRefresh = findViewById(R.id.btn_refresh);
        btnFlush = findViewById(R.id.btn_flush);
        tvTotalPlayed = findViewById(R.id.tv_total_played);
        tvMaxDuration = findViewById(R.id.tv_max_duration);
        tvMaxDuration.setText(durationToText(volumeControlService.getMaxUnsafeMusicPlayDuration()));

        updateProgressBar();

        btnRefresh.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateProgressBar();
            }
        });

        btnFlush.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                volumeControlService.putTotalUnsafeMilliseconds(MainActivity.this.getContentResolver(), 1);
                updateProgressBar();
            }
        });

        ReminderServiceScheduler reminderScheduler = new ReminderServiceScheduler(getApplicationContext(), 0.1);
        //TODO: load hours from preferences
        reminderScheduler.scheduleReminderService(8);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case R.id.action_settings:
                Intent startSettingsActivityIntent = SettingsActivity.makeStartIntent(this);
                startActivity(startSettingsActivityIntent);
                break;
        }

        return super.onOptionsItemSelected(item);
    }


    private void updateProgressBar() {
        try {
            int currentState = volumeControlService.getUnsafeMilliseconds(getContentResolver());
            tvTotalPlayed.setText(durationToText(currentState));
            progressCurrent.setMax(volumeControlService.getMaxUnsafeMusicPlayDuration());
            progressCurrent.setProgress(currentState);

        } catch (Settings.SettingNotFoundException e) {
            Log.e(TAG, "Could not load setting", e);
            Toast.makeText(this,"Could not load current value.", Toast.LENGTH_SHORT).show();
        }
    }

    private String durationToText(int duration) {
        long hours = TimeUnit.MILLISECONDS.toHours(duration);
        long minutes = TimeUnit.MILLISECONDS.toMinutes(duration) - hours * 60;

        return String.format("%02d:%02d", hours, minutes);
    }
}

File: app/src/main/java/com/trueapps/volumecontrol/MainActivity.kt
package com.trueapps.volumecontrol

import android.os.Bundle
import android.provider.Settings.SettingNotFoundException
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.trueapps.volumecontrol.settings.SettingsActivity
import java.util.concurrent.TimeUnit


class MainActivity : AppCompatActivity() {
    private lateinit var progressCurrent: ProgressBar
    private lateinit var btnReset: Button
    private lateinit var tvRemainingTimeDisplay: TextView
    private lateinit var tvRemainingTime: TextView
    private lateinit var tvMaxDuration: TextView

    private lateinit var secureSettings: SecureSettingsManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        secureSettings = SecureSettingsManager()

        progressCurrent = findViewById(R.id.progress_current)
        progressCurrent.rotation = 180f
        btnReset = findViewById(R.id.btn_flush)
        tvRemainingTimeDisplay = findViewById(R.id.tv_remaining_time_display)
        tvRemainingTime = findViewById(R.id.tv_remaining_time_small)
        tvMaxDuration = findViewById<TextView>(R.id.tv_max_duration).apply {
            text = durationToText(secureSettings.unsafeVolumeMusicActiveMsMax, R.string.duration_pattern)
        }

        btnReset.setOnClickListener {
            resetUnsafeMillis()
            updateProgressBar()
            showRebootReminderDialog()
        }
    }

    private fun resetUnsafeMillis() {
        try {
            secureSettings.updateTotalUnsafeMilliseconds(this@MainActivity.contentResolver, 1)
        } catch (e: SecurityException) {
            Log.e(TAG, "Security exception: no permissions to write secure settings", e)
            showPermissionRequiredDialog()
        }
    }

    override fun onResume() {
        super.onResume()
        updateProgressBar()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                val startSettingsActivityIntent = SettingsActivity.makeStartIntent(this)
                startActivity(startSettingsActivityIntent)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun updateProgressBar() {
        try {
            val remainingTime = secureSettings.unsafeVolumeMusicActiveMsMax - secureSettings.readUnsafeMilliseconds(contentResolver)
            tvRemainingTimeDisplay.text = durationToText(remainingTime, R.string.display_duration_pattern)
            tvRemainingTime.text = durationToText(remainingTime, R.string.duration_pattern)
            progressCurrent.max = secureSettings.unsafeVolumeMusicActiveMsMax
            progressCurrent.progress = remainingTime
        } catch (e: SettingNotFoundException) {
            Log.e(TAG, "Could not load setting", e)
            Toast.makeText(this, "Could not load current value.", Toast.LENGTH_SHORT).show()
        }
    }

    private fun showRebootReminderDialog() {
        AlertDialog.Builder(this)
                .setTitle(R.string.attention)
                .setMessage(R.string.remind_to_reboot_message)
                .setPositiveButton(R.string.ok) { dialog, id -> }
                .show()
    }

    private fun showPermissionRequiredDialog() {
        AlertDialog.Builder(this)
                .setTitle(R.string.no_permission)
                .setMessage(getString(R.string.grant_permission_message))
                .setPositiveButton(R.string.ok) { dialog, id -> }
                .show()
    }

    private fun durationToText(duration: Int, patternResId: Int): String {
        val hours = TimeUnit.MILLISECONDS.toHours(duration.toLong())
        val minutes = TimeUnit.MILLISECONDS.toMinutes(duration.toLong()) - hours * 60
        return String.format(getString(patternResId), hours, minutes)
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}